generic/tests/pktgen_perf.py

```diff
@@ -0,0 +1,196 @@
+import logging
+import os
+import re
+import aexpect
+
+from avocado.utils import process
+
+from virttest import remote
+from virttest import data_dir
+from virttest import utils_net
+from virttest import utils_test
+from virttest import utils_misc
+from virttest import error_context
+
+
+def format_result(result, base="12", fbase="2"):
+    """
+    Format the result to a fixed length string.
+
+    :param result: result need to convert
+    :param base: the length of converted string
+    :param fbase: the decimal digit for float
+    """
+    if isinstance(result, str):
+        value = "%" + base + "s"
+    elif isinstance(result, int):
+        value = "%" + base + "d"
+    elif isinstance(result, float):
+        value = "%" + base + "." + fbase + "f"
+    return value % result
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Run Pktgen test between host/guest
+
+    1) Boot the main vm, or just grab it if it's already booted.
+    2) Configure pktgen on guest or host
+    3) Run pktgen test, finish when timeout
+
+    :param test: QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env: Dictionary with test environment.
+    """
+
+    def _pin_vm_threads(node):
+        """
+        pin guest vcpu and vhost threads to cpus of a numa node repectively
+
+        :param node: which numa node to pin
+        """
+        if node:
+            if not isinstance(node, utils_misc.NumaNode):
+                node = utils_misc.NumaNode(int(node))
+            utils_test.qemu.pin_vm_threads(vm, node)
+
+    login_timeout = float(params.get("login_timeout", 360))
+    timeout = float(params.get("pktgen_test_timeout", "240"))
+    run_threads = params.get("pktgen_threads", 1)
+    record_list = params.get("record_list")
+    error_context.context("Init the VM, and try to login", logging.info)
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=login_timeout)
+    session_serial = vm.wait_for_serial_login(timeout=timeout)
+
+    # print numa information on host and pinning vhost and vcpus to cpus
+    process.system_output("numactl --hardware")
+    process.system_output("numactl --show")
+    _pin_vm_threads(params.get("numa_node"))
+
+    # get parameter from dictionary
+    pktgen_stress_timeout = float(params.get("pktgen_test_timeout", 60))
+    shell_port = params.get("shell_port", "22")
+    username = params.get("username", "root")
+    password = params.get("password", "123456")
```

Comment:
Pls specify the default password 'kvmautotest'

Reply:
Ｕpdated.